pkg/member/member_add.go

```diff
@@ -0,0 +1,264 @@
+package member
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/gardener/etcd-backup-restore/pkg/etcdutil"
+	"github.com/gardener/etcd-backup-restore/pkg/etcdutil/client"
+	"github.com/gardener/etcd-backup-restore/pkg/miscellaneous"
+	brtypes "github.com/gardener/etcd-backup-restore/pkg/types"
+	"github.com/sirupsen/logrus"
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/etcdserver/api/v3rpc/rpctypes"
+	"gopkg.in/yaml.v2"
+	"k8s.io/client-go/util/retry"
+)
+
+const (
+	// retryPeriod is the peroid after which an operation is retried
+	retryPeriod = 5 * time.Second
+
+	// etcdTimeout is timeout for etcd operations
+	etcdTimeout = 10 * time.Second
+)
+
+// AddMember interface defines the functionalities needed to add a member to the etcd cluster
+type AddMember interface {
+	// AddMemberAsLearner add a new member as a learner to the etcd cluster
+	AddMemberAsLearner(*logrus.Logger, *brtypes.EtcdConnectionConfig)
+
+	// IsMemberInCluster checks is the current members peer URL is already part of the etcd cluster
+	IsMemberInCluster(*logrus.Logger, *brtypes.EtcdConnectionConfig)
+
+	// PromoteMember promotes an etcd member from a learner to a voting member of the cluster. This will succeed only if its logs are caught up with the leader
```

Comment:
```suggestion
	// PromoteMember promotes an etcd member from a learner to a voting member of the cluster. This will succeed if and only if:  learner is in a healthy state and the learner is in sync with leader.
```